evm/src/lib.rs

```diff
@@ -1,3 +1,153 @@
+//! let (proof_3, pv_3) =
+//!     prover_state.prove_root(all_stark, config, inputs_2, &mut timing, None);
+//!
+//! // Now aggregate proofs for txn 1 and 2.
+//! let (agg_proof_1_2, pv_1_2) =
+//!     prover_state.prove_aggreg(false, proof_1, pv_1, false, proof_2, pv_2);
+//!
+//! // Now aggregate the newly generated aggregation proof with the last regular txn proof.
+//! let (agg_proof_1_3, pv_1_3) =
+//!     prover_state.prove_aggreg(true, agg_proof_1_2, pv_1_2, false, proof_3, pv_3);
+//! ```
+//!
+//! **Note**: The proofs provided to the [prove_aggreg](AllRecursiveCircuits::prove_aggreg) method *MUST* have contiguous states.
+//! Trying to combine `proof_1` and `proof_3` from the example above would fail.
+//!
+//! ## Block proofs
+//!
+//! Once all transactions of a block have been proven and we are left with a single aggregation proof and its public values,
+//! we can then wrap it into a final block proof, attesting validity of the entire block.
+//! This [prove_block](AllRecursiveCircuits::prove_aggreg) method accepts an optional previous block proof as argument,
+//! which will then try combining the previously proven block with the current one, generating a validity proof for both.
+//! Applying this process from genesis would yield a single proof attesting correctness of the entire chain.
+//!
+//! ```ignore
+//! let previous_block_proof = { ... };
+//! let (block_proof, block_public_values) =
+//!     prover_state.prove_block(Some(&previous_block_proof), &agg_proof, agg_pv)?;
+//! ```
+//!
+//! ### Checkpoint heights
+//!
+//! Because the process of always providing a previous block proof when generating a proof for the current block may yield some
+//! undesirable issues, the plonky2 zk-EVM supports checkpoint heights. At given block heights, the prover does not have to pass
+//! a previous block proof. This would in practice correspond to block heights at which a proof has been generated and sent to L1
+//! for settlement.
+//!
+//! # Prover state serialization
+//!
+//! Because the recursive circuits only need to be generated once, they can be saved to disk once the preprocessing phase
+//! completed successfully, and deserialized on-demand.
+//! The plonky2 zk-EVM provides serialization methods to convert the entire prover state to a vector of bytes, and vice-versa.
+//! This requires the use of custom serializers for gates and generators for proper recursive circuit encoding. This crate provides
+//! default serializers supporting all custom gates and associated generators defined within the [`plonky2`] crate.
+//!
+//! ```ignore
+//! let prover_state = AllRecursiveCircuits::<F, C, D>::new(...);
+//!
+//! // Default serializers
+//! let gate_serializer = DefaultGateSerializer;
+//! let generator_serializer = DefaultGeneratorSerializer::<C, D> {
+//!     _phantom: PhantomData::<C>,
+//! };
+//!
+//! // Serializes the prover state to a sequence of bytes
```

Comment:
Can you provide an order of magnitude for how big the proofs are when serialized?

Reply:
the proofs or the prover state?

Reply:
The prover state will range between 20GB to 25GB when serialized.
The proofs are in the ballpark of ~150kB, regardless of the aggregation level, based on the current config (although the initial STARK proofs for a transaction, depend on the actual transaction execution, and may be several MBs large).

Reply:
prover state! I'm picturing some poor devops person that is trying to figure out how to set up a proving machine. Could be useful to have a hunch of how much disk space is actually needed.

Reply:
The alternative is to split the prover state (it's been rendered possible with #1438), by loading only upper circuits needed after shrinking the STARK proofs and split the transaction proof process into three chunks:

- compute the initial STARK proofs
- load the related recursive table circuits
- finish the txn proof (basically what's in `prove_root` after the first line)

But this is out of scope of the plonky2_evm lib.